libcodechecker/server/product_db_access_handler.py

```diff
@@ -0,0 +1,300 @@
+        self.__session = config_session
+        self.__product = routed_product
+
+    @timeit
+    def getAPIVersion(self):
+        return shared_constants.API_VERSION
+
+    @timeit
+    def getPackageVersion(self):
+        return self.__package_version
+
+    @timeit
+    def getProducts(self, product_endpoint_filter, product_name_filter):
+        """
+        Get the list of products configured on the server.
+        """
+
+        result = []
+
+        try:
+            prods = self.__session.query(Product)
+
+            if product_endpoint_filter:
+                prods = prods.filter(Product.endpoint.like(
+                    '%{0}%'.format(product_endpoint_filter)))
+
+            if product_name_filter:
+                prods = prods.filter(Product.display_name.like(
+                    '%{0}%'.format(product_name_filter)))
+
+            prods = prods.all()
+
+            for prod in prods:
+                server_product = self.__server.get_product(prod.endpoint)
+
+                if not server_product:
+                    # TODO: Better support this, if the configuration database
+                    # is mounted to multiple servers?
+                    LOG.error("Product '{0}' was found in the configuration "
+                              "database, but no database connection is "
+                              "present. Was the configuration database "
+                              "connected to multiple servers?"
+                              .format(prod.endpoint))
+                    LOG.info("Please restart the server to make this "
+                             "product available.")
+                    continue
+
+                # Clients are expected to use this method to query if
+                # the product exists and usable. Usability sometimes requires
+                # "updating" the "is connected" status of the database.
+                if not server_product.connected:
+                    server_product.connect()
+
+                name = base64.b64encode(prod.display_name.encode('utf-8')) \
+                    if prod.display_name \
+                    else base64.b64encode(prod.endpoint.encode('utf-8'))
+                desc = base64.b64encode(prod.description.encode('utf-8')) \
+                    if prod.description else None
+
+                result.append(ttypes.Product(
+                    id=prod.id,
+                    endpoint=prod.endpoint,
+                    displayedName_b64=name,
+                    description_b64=desc,
+                    connected=server_product.connected,
+                    accessible=True))
+
+            return result
+
+        except sqlalchemy.exc.SQLAlchemyError as alchemy_ex:
+            msg = str(alchemy_ex)
+            LOG.error(msg)
+            raise shared.ttypes.RequestFailed(shared.ttypes.ErrorCode.DATABASE,
+                                              msg)
+
+    @timeit
+    def getCurrentProduct(self):
```

Comment:
The variable pointed by `self.__product` has a global lifetime of the entire server, as this is a `Product` connection object, containing the database engine and everything, defined in `client_db_access_server`. For each product currently connected, such an object exists.

`self` itself is the Thrift handler, which is constructed per-API-call in `do_POST` of the server handler, which gives one of these `Product` instances as argument.